### PR TITLE
 Checkpoints are always disabled (#1315)

### PR DESCRIPTION
**Pull request by Neurosploit**
Fix for https://github.com/stratisproject/StratisBitcoinFullNode/issues/1315 and unit tested the whole class.

### CODE REVIEW

File: src/Stratis.Bitcoin.Features.Consensus/Checkpoints.cs [hunk 1]
```diff
@@ -134,46 +134,54 @@ public class Checkpoints : ICheckpoints
         };
 
         /// <summary>Checkpoints for the specific instance of the class and its network.</summary>
-        private readonly Dictionary<int, CheckpointInfo> checkpoints;
+        private readonly Dictionary<int, CheckpointInfo> networkSpecificCheckpoints;
+
+        /// <summary>Empty list of checkpoints for the specific instance of the class in case the use of checkpoints is disabled in the consensus settings.</summary>
+        private readonly Dictionary<int, CheckpointInfo> emptyCheckpoints = new Dictionary<int, CheckpointInfo>();
```
Comment (bokobza) [inline src/Stratis.Bitcoin.Features.Consensus/Checkpoints.cs]: We usually avoid instantiating objects outside of the ctor.
Also it seems this object is never filled (empty, I get it), and then only returned in `GetCheckpoints()`. Is it really necessary to have?

Comment (Neurosploit) [reply src/Stratis.Bitcoin.Features.Consensus/Checkpoints.cs]: I could just return a new empty one instead.

File: src/Stratis.Bitcoin.Features.Consensus/Checkpoints.cs [hunk 2]
```diff
@@ -182,8 +190,16 @@ public bool CheckHardened(int height, uint256 hash)
...
         {
             CheckpointInfo checkpoint;
-            this.checkpoints.TryGetValue(height, out checkpoint);
+            this.GetCheckpoints().TryGetValue(height, out checkpoint);
             return checkpoint;
         }
+
+        private Dictionary<int, CheckpointInfo> GetCheckpoints()
+        {
+            if (this.ConsensusSettings == null)
```
Comment (bokobza) [inline src/Stratis.Bitcoin.Features.Consensus/Checkpoints.cs]:
```
if (this.ConsensusSettings == null || !this.ConsensusSettings.UseCheckpoints)
    return this.emptyCheckpoints;

return this.networkSpecificCheckpoints;
```
seems more condensed to me.

Comment (Neurosploit) [reply src/Stratis.Bitcoin.Features.Consensus/Checkpoints.cs]: Yep

**Comment by noescape00** [comment]
Yes, sure. Will take a look in couple of hours.

**Comment by dangershony** [comment]
Looks good to me.
@Neurosploit eventually the checkpoints themselves need to go to the network file 
If you are up for it to do another PR after this to move checkpoints to Network and add them to the network builder.

File: src/Stratis.Bitcoin.Features.Consensus/Checkpoints.cs [hunk 3]
```diff
@@ -182,8 +187,16 @@ public bool CheckHardened(int height, uint256 hash)
...
         {
             CheckpointInfo checkpoint;
-            this.checkpoints.TryGetValue(height, out checkpoint);
+            this.GetCheckpoints().TryGetValue(height, out checkpoint);
             return checkpoint;
         }
+
+        private Dictionary<int, CheckpointInfo> GetCheckpoints()
+        {
+            if (this.ConsensusSettings == null || !this.ConsensusSettings.UseCheckpoints)
```
Comment (noescape00) [inline src/Stratis.Bitcoin.Features.Consensus/Checkpoints.cs]: Can the `ConsensusSettings ` be null here?

**Comment by noescape00** [comment]
Reviewed it, looks nice!

File: src/Stratis.Bitcoin.Features.Consensus/Checkpoints.cs (hunk 3 above)
Comment (Neurosploit) [reply src/Stratis.Bitcoin.Features.Consensus/Checkpoints.cs]: Yes it can if you call the parameterless constructor.

**Pull request merged**